camel/agents/insight_agent.py

```diff
@@ -0,0 +1,313 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import re
+from typing import Any, Dict, Optional, Union
+
+from camel.agents import ChatAgent
+from camel.messages import BaseMessage
+from camel.prompts import TextPrompt
+from camel.types import ModelType, RoleType
+
+
+class InsightAgent(ChatAgent):
+    r"""An agent that aims to generate insights from a provided text.
+    Args:
+        model_type (ModelType, optional): The type of model to use for the
+            agent. (default: :obj:`ModelType.GPT_3_5_TURBO`)
+        model_config (Any, optional): The configuration for the model.
+            (default: :obj:`None`)
+    """
+
+    def __init__(
+        self,
+        model_type: ModelType = ModelType.GPT_3_5_TURBO,
+        model_config: Optional[Any] = None,
+    ) -> None:
+        system_message = BaseMessage(
+            role_name="Insight Agent",
+            role_type=RoleType.ASSISTANT,
+            meta_dict=None,
+            content="You generate insights from a provided text.",
+        )
+        super().__init__(system_message, model_type, model_config)
+
+    def get_insights(
+        self,
+        context_text: Union[str, TextPrompt],
+        insights_instruction: Optional[Union[str, TextPrompt]] = None,
+    ) -> Dict[str, Dict[str, str]]:
+        r"""Generate insights from the context text. And the function
+        pulls as many meaningful insights as possible, organizing them
+        according to the "RULES OF INSIGHTS GENERATION" and the "ANSWER
+        TEMPLATE":
+            - Topic Segmentation
+            - Entity Recognition
+            - Extract Details
+            - Contextual Understanding
+            - Formulate Questions
+            - Answer to "Formulate Questions" using CONTEXT TEXT
+            - Iterative Feedback
+
+        Args:
+            context_text (Union[str, TextPrompt]): The context text to
+                generate insights from.
+            insights_instruction (Optional[Union[str, TextPrompt]], optional):
+                The instruction for generating insights. (default: :obj:`None`)
+
+        Returns:
+            Dict[str, Dict[str, str]]: The generated insights from the context
+                text.
+        """
+        self.reset()
+
+        insights_instruction_prompt = \
```

Comment:
Maybe we can move these prompts to somewhere else to make this cleaner?

Reply:
Good idea, but the insight agent is a single agent who doesn't belong to any agent cluster. Will this disrupt the original clear file structure, if I move the prompt to the file 'prompts'? What do you think of it?

Reply:
I think current get_insights looks good

Reply:
Maybe we can move to the top of the file
```python3
INSIGHTS_INSTRUCTION_PROMPT = ...
```